### PR TITLE
[dotnet] UIMain changes

### PR DESCRIPTION
**Pull request by stephen-hawley**
handling of string array for args

### CODE REVIEW

File: src/ObjCRuntime/TransientString.cs [hunk 1]
```diff
@@ -85,5 +85,26 @@ public void Dispose ()
...
 			return result;
 		}
+
+		public static IntPtr AllocStringArray (string []? arr, Encoding encoding = Encoding.Auto)
+		{
+			if (arr is null)
+				return IntPtr.Zero;
+			var ptrArr = Marshal.AllocHGlobal (arr.Length * IntPtr.Size);
+			for (int i = 0; i < arr.Length; i++) {
+				Marshal.WriteIntPtr (ptrArr + 4 * i, new TransientString (arr [i], encoding));
```
Comment (rolfbjarne) [inline src/ObjCRuntime/TransientString.cs:95]:
```suggestion
				Marshal.WriteIntPtr (ptrArr + IntPtr.Size * i, new TransientString (arr [i], encoding));
```

File: src/ObjCRuntime/TransientString.cs [hunk 2]
```diff
@@ -85,5 +85,26 @@ public void Dispose ()
...
+			}
+			return ptrArr;
+		}
+
+		public static void FreeStringArray (IntPtr ptrArr, int count)
+		{
+			if (ptrArr == IntPtr.Zero)
+				return;
+			for (int i = 0; i < count; i++) {
+				Marshal.FreeHGlobal (Marshal.ReadIntPtr (ptrArr + 4 * i));
```
Comment (rolfbjarne) [inline src/ObjCRuntime/TransientString.cs:105]:
```suggestion
				Marshal.FreeHGlobal (Marshal.ReadIntPtr (ptrArr + IntPtr.Size * i));
```

File: src/UIKit/UIApplication.cs [hunk 3]
```diff
@@ -50,11 +50,21 @@ partial class UIApplication
...
+#else
 		extern static int xamarin_UIApplicationMain (int argc, /* char[]* */ string []? argv, /* NSString* */ IntPtr principalClassName, /* NSString* */ IntPtr delegateClassName, out IntPtr gchandle);
+#endif
 
 		static int UIApplicationMain (int argc, /* char[]* */ string []? argv, /* NSString* */ IntPtr principalClassName, /* NSString* */ IntPtr delegateClassName)
 		{
+#if NET
+			var strArr = TransientString.AllocStringArray (argv);
+			var rv = xamarin_UIApplicationMain (argc, strArr, principalClassName, delegateClassName, out var gchandle);
+			TransientString.FreeStringArray (strArr, argc);
```
Comment (rolfbjarne) [inline src/UIKit/UIApplication.cs:64]: This makes sure we don't crash if `argc` is wrong:

```suggestion
			TransientString.FreeStringArray (strArr, argv?.Length ?? 0);
```

File: src/UIKit/UIApplication.cs [hunk 4]
```diff
@@ -50,11 +50,21 @@ partial class UIApplication
 		// We link with __Internal here so that this function is interposable from third-party native libraries.
 		// See: https://github.com/xamarin/MicrosoftInTune/issues/3 for an example.
 		[DllImport ("__Internal")]
+#if NET
+		extern static int xamarin_UIApplicationMain (int argc, /* char[]* */ IntPtr argv, /* NSString* */ IntPtr principalClassName, /* NSString* */ IntPtr delegateClassName, out IntPtr gchandle);
```
Comment (rolfbjarne) [inline src/UIKit/UIApplication.cs:54]: Is there any reason we can't always use the new code (i.e. don't check for NET)?

**Review by rolfbjarne — CHANGES_REQUESTED** [review]

**Review by rolfbjarne — APPROVED** [review]

**Review by dalexsoto — APPROVED** [review]

**Review by haritha-mohan — APPROVED** [review]

**Pull request merged**